### PR TITLE
Hierarchical symbols

### PR DESCRIPTION
**Pull request by gatesn**
@lgeiger want to decide on what the symbols look like:

```
import sys

a = 'hello'

class B:
    def __init__(self):
        x = 2
        self.y = x

def main(x):
    y = 2 * x
    return y
```

IMO:

- `sys`
- `a`
- `B`
- `B.__init__`
- `B.__init__.x`
- `main`
- `main.y`

### CODE REVIEW

**Comment by lgeiger** [comment]
Thanks for continuing the efforts!

This is how it looks in Atom:
bildschirmfoto 2017-10-28 um 20 53 35

The filename is `jedie.py`. [This line](https://github.com/palantir/python-language-server/pull/175/files#diff-cee448b8821a222944d7e6f2ca332ddaR40) used to prevent it from always showing.

Personally I'd prefer to not show `x` any `y` per default because things may become pretty cluttered on large files, but I'm happy to go with this and let more people chime in on the discussion.
I think it's OK to go with this for now and see if people complain about `pyls` showing too many symbols. Since `pyls` is pre 1.0 we can always iterate later.

`pyls/plugins/symbols.py` looks like this:
bildschirmfoto 2017-10-28 um 21 02 45

File: pyls/plugins/symbols.py [hunk 1]
```diff
@@ -9,42 +9,40 @@
...
-         (parent and parent.parent()))) or name.type == 'param':
-        return None
-    else:
-        result['name'] = name.name
-
-    result['location'] = {'uri': uri, 'range': _range(name)}
-    result['kind'] = _kind(name)
-
-    # All assignments have the filename as a there toplevel parent.
-    if parent and parent.parent():
```
Comment (lgeiger) [inline pyls/plugins/symbols.py]: We still need this check, otherwise the filename is always the top most symbol.

**Comment by gatesn** [comment]
I added the check to remove the top-level module name. But there's not sensible way to get rid of x and y without excluding things like a. So I'm going to leave it in for now.

**Comment by gatesn** [comment]
@lgeiger I just had a thought... the names should be fully qualified. Otherwise a variable defined in constructors of two classes (__init__ method) can't be disambiguated.

**Comment by lgeiger** [comment]
> I added the check to remove the top-level module name.

👍 

>  But there's not sensible way to get rid of x and y without excluding things like a. So I'm going to leave it in for now.

That's totally possible. That's what [this check did](https://github.com/palantir/python-language-server/pull/176/commits/55913e936e60b732aea05d301e9344d391eb42c8#diff-cee448b8821a222944d7e6f2ca332ddaR31).


> @lgeiger I just had a thought... the names should be fully qualified. Otherwise a variable defined in constructors of two classes (init method) can't be disambiguated.

I don't know if I fully understand you, but I think they can be distinguished easily by the indentation in the UI since the `__init__` methods hase different parents.

**Comment by gatesn** [comment]
What I mean is that `a` will have containerName `__init__`, as will `b`.

**Comment by gatesn** [comment]
I think we already make that check right? https://github.com/palantir/python-language-server/pull/176/files#diff-cee448b8821a222944d7e6f2ca332ddaR28

[edit] I see what's missing. Let me fix.

**Comment by lgeiger** [comment]
> What I mean is that a will have containerName __init__, as will b.

Yes, that's why the correct start and end positions are important for the UI to figure out where they belong.

**Comment by gatesn** [comment]
Got it, the positions are used in addition to containerName. 

Also, on second thoughts, a variable local to a function is still a symbol. So going to leave them.

**Comment by lgeiger** [comment]
> I think we already make that check right?

No, the check [here](https://github.com/palantir/python-language-server/pull/176/commits/55913e936e60b732aea05d301e9344d391eb42c8#diff-cee448b8821a222944d7e6f2ca332ddaR31) also tests if the `statement` is defined inside a function and omits it. If `parent.parent()` isn't defined it's in the global scope, otherwise it was defined inside a function.

> Got it, the positions are used in addition to containerName.

Yeah, it's not exactly the most elegant way to infer hierarchy 😄 

> Also, on second thoughts, a variable local to a function is still a symbol. So going to leave them.

I'm OK with this. I don't know what the canonical way would be.

**Comment by gatesn** [comment]
Ok, sounds sounds like we're in agreement. I also don't mind if the client excludes variables whose container symbol is a function. Most outline views tend to be expandable trees though, so I wonder if it can just be solved with default level of expanded subtrees.

Anyway, going to merge to get this out the door. Thanks for all your help / opinions!

**Comment by lgeiger** [comment]
> so I wonder if it can just be solved with default level of expanded subtrees.

It's not so easy to solve this with default levels since one would also exclude class methods then.

> Anyway, going to merge to get this out the door. Thanks for all your help / opinions!

Cool looking forward to the next release.

**Comment by bhack** [comment]
Do you think that we will have a new tagged release for this soon?

**Comment by bhack** [comment]
@lgeiger The is a new release now.

**Pull request merged**